File: paper-roll-pwa/src/utils/validators.ts
export function isValidINN(inn: string): boolean {
  if (!/^\d+$/.test(inn)) {
    return false;
  }

  if (inn.length === 10) {
    const coefficients = [2, 4, 10, 3, 5, 9, 4, 6, 8];
    let sum = 0;
    for (let i = 0; i < 9; i++) {
      sum += parseInt(inn[i]) * coefficients[i];
    }
    let check = sum % 11;
    if (check > 9) check = 0;
    return check === parseInt(inn[9]);
  }

  if (inn.length === 12) {
    const coef1 = [7, 2, 4, 10, 3, 5, 9, 4, 6, 8];
    let sum1 = 0;
    for (let i = 0; i < 10; i++) {
      sum1 += parseInt(inn[i]) * coef1[i];
    }
    let check1 = sum1 % 11;
    if (check1 > 9) check1 = 0;

    const coef2 = [3, 7, 2, 4, 10, 3, 5, 9, 4, 6, 8];
    let sum2 = 0;
    for (let i = 0; i < 11; i++) {
      sum2 += parseInt(inn[i]) * coef2[i];
    }
    let check2 = sum2 % 11;
    if (check2 > 9) check2 = 0;

    return check1 === parseInt(inn[10]) && check2 === parseInt(inn[11]);
  }

  return false;
}

File: t_bot/utils/validators.py
def is_valid_inn(inn: str) -> bool:
    if not inn.isdigit():
        return False
    
    if len(inn) == 10:  # Юрлицо
        coefficients = [2, 4, 10, 3, 5, 9, 4, 6, 8]
        n = sum(int(inn[i]) * coefficients[i] for i in range(9))
        check = n % 11
        if check > 9:
            check = 0
        return check == int(inn[9])
        
    elif len(inn) == 12:  # ИП
        coefficients_1 = [7, 2, 4, 10, 3, 5, 9, 4, 6, 8]
        n1 = sum(int(inn[i]) * coefficients_1[i] for i in range(10))
        check1 = n1 % 11
        if check1 > 9:
            check1 = 0
            
        coefficients_2 = [3, 7, 2, 4, 10, 3, 5, 9, 4, 6, 8]
        n2 = sum(int(inn[i]) * coefficients_2[i] for i in range(11))
        check2 = n2 % 11
        if check2 > 9:
            check2 = 0
            
        return check1 == int(inn[10]) and check2 == int(inn[11])
    
    return False